<a href="https://colab.research.google.com/github/PranayUmredkar97/Demand-Predicition-for-Public-Transport/blob/main/Copy_of_Demand_Prediction_for_Public_Transport_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b><u> Project Title : Traffic Jam: Predicting People's Movement into Nairobi </u></b>

## <b> Problem Description </b>

### This challenge asks you to build a model that predicts the number of seats that Mobiticket can expect to sell for each ride, i.e. for a specific route on a specific date and time. There are 14 routes in this dataset. All of the routes end in Nairobi and originate in towns to the North-West of Nairobi towards Lake Victoria.


### The towns from which these routes originate are:

* Awendo
* Homa Bay
* Kehancha
* Kendu Bay
* Keroka
* Keumbu
* Kijauri
* Kisii
* Mbita
* Migori
* Ndhiwa
* Nyachenge
* Oyugis
* Rodi
* Rongo
* Sirare
* Sori

### The routes from these 14 origins to the first stop in the outskirts of Nairobi takes approximately 8 to 9 hours from time of departure. From the first stop in the outskirts of Nairobi into the main bus terminal, where most passengers get off, in Central Business District, takes another 2 to 3 hours depending on traffic.

### The three stops that all these routes make in Nairobi (in order) are:

1. Kawangware: the first stop in the outskirts of Nairobi
2. Westlands
3. Afya Centre: the main bus terminal where most passengers disembark

### All of these points are mapped [here](https://www.google.com/maps/d/viewer?mid=1Ef2pFdP8keVHHid8bwju2raoRvjOGagN&ll=-0.8281897101491997%2C35.51706279999996&z=8).

### Passengers of these bus (or shuttle) rides are affected by Nairobi traffic not only during their ride into the city, but from there they must continue their journey to their final destination in Nairobi wherever that may be. Traffic can act as a deterrent for those who have the option to avoid buses that arrive in Nairobi during peak traffic hours. On the other hand, traffic may be an indication for people’s movement patterns, reflecting business hours, cultural events, political events, and holidays.

# **Import Packages**

In [36]:
#imporgt required packages and libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go
import plotly.express as px
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [2]:
#set style for graph
sns.set_style(style='darkgrid')

# Setting fontsizes,fontweight,label weight for labels,titles.
plt.rcParams["font.weight"] = "bold"
plt.rcParams["axes.labelweight"] = "bold"
plt.rcParams["axes.titlesize"] = 25
plt.rcParams["axes.titleweight"] = 'bold'
plt.rcParams['xtick.labelsize']=15
plt.rcParams['ytick.labelsize']=15
plt.rcParams["axes.labelsize"] = 20
plt.rcParams["legend.fontsize"] = 15
plt.rcParams["legend.title_fontsize"] = 15

# **Data Preperation**

## **Mounting Drive**

In [3]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [4]:
#csv file location
path = '/content/drive/MyDrive/machine learning regression nairobi/train_revised.csv'

## **Reading data in notebook**

In [5]:
df = pd.read_csv(path)

## **Data Inspection**

In [6]:
#first look checking top 5 rows
df.head()

ride_id seat_number payment_method payment_receipt travel_date travel_time  \
0     1442         15A          Mpesa      UZUEHCBUSO    17-10-17        7:15   
1     5437         14A          Mpesa      TIHLBUSGTE    19-11-17        7:12   
2     5710          8B          Mpesa      EQX8Q5G19O    26-11-17        7:05   
3     5777         19A          Mpesa      SGP18CL0ME    27-11-17        7:10   
4     5778         11A          Mpesa      BM97HFRGL9    27-11-17        7:12   

  travel_from travel_to car_type  max_capacity  
0      Migori   Nairobi      Bus            49  
1      Migori   Nairobi      Bus            49  
2      Keroka   Nairobi      Bus            49  
3    Homa Bay   Nairobi      Bus            49  
4      Migori   Nairobi      Bus            49

In [7]:
df.tail()

ride_id seat_number payment_method payment_receipt travel_date  \
51640    13826          9B          Mpesa      8V2XDDZR6V    20-04-18   
51641    13809         18A          Mpesa      4PEBSVJSNK    20-04-18   
51642    13809         17A          Mpesa      LVN64LZDNN    20-04-18   
51643    13796         16B          Mpesa      REYBSKTYWN    20-04-18   
51644    14304           7          Mpesa      AQN7FBUSGP    14-11-17   

      travel_time travel_from travel_to car_type  max_capacity  
51640        8:00      Awendo   Nairobi      Bus            49  
51641        8:00      Migori   Nairobi      Bus            49  
51642        8:00      Migori   Nairobi      Bus            49  
51643        7:08      Awendo   Nairobi      Bus            49  
51644        5:10       Kisii   Nairobi      Bus            49

In [8]:
#check information about the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51645 entries, 0 to 51644
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   ride_id          51645 non-null  int64 
 1   seat_number      51645 non-null  object
 2   payment_method   51645 non-null  object
 3   payment_receipt  51645 non-null  object
 4   travel_date      51645 non-null  object
 5   travel_time      51645 non-null  object
 6   travel_from      51645 non-null  object
 7   travel_to        51645 non-null  object
 8   car_type         51645 non-null  object
 9   max_capacity     51645 non-null  int64 
dtypes: int64(2), object(8)
memory usage: 3.9+ MB


In [9]:
df.describe(include='all').T

count unique         top   freq          mean          std  \
ride_id          51645.0    NaN         NaN    NaN  10188.645793  2211.295708   
seat_number        51645     61           1   2065           NaN          NaN   
payment_method     51645      2       Mpesa  51532           NaN          NaN   
payment_receipt    51645  51645  UZUEHCBUSO      1           NaN          NaN   
travel_date        51645    149    10-12-17    856           NaN          NaN   
travel_time        51645     78        7:09   3926           NaN          NaN   
travel_from        51645     17       Kisii  22607           NaN          NaN   
travel_to          51645      1     Nairobi  51645           NaN          NaN   
car_type           51645      2         Bus  31985           NaN          NaN   
max_capacity     51645.0    NaN         NaN    NaN     34.534321    18.451193   

                    min     25%      50%      75%      max  
ride_id          1442.0  8287.0  10353.0  12029.0  20117.0  
seat_number         NaN     NaN      NaN      NaN      NaN  
payment_method      NaN     NaN      NaN      NaN      NaN  
payment_receipt     NaN     NaN      NaN      NaN      NaN  
travel_date         NaN     NaN      NaN      NaN      NaN  
travel_time         NaN     NaN      NaN      NaN      NaN  
travel_from         NaN     NaN      NaN      NaN      NaN  
travel_to           NaN     NaN      NaN      NaN      NaN  
car_type            NaN     NaN      NaN      NaN      NaN  
max_capacity       11.0    11.0     49.0     49.0     49.0

## Inference


*   There are 61 unique seat numbers in this dataset.
*   There are two payment method used by travellers and most of them used Mpesa for payment.
* We have data for 149 days in our dataset.
* There are two different types of car used for travelling and mostly bus having max capcity of 49 and other is shuttle having max capacity of 11.


In [10]:
#check columns names
df.columns

Index(['ride_id', 'seat_number', 'payment_method', 'payment_receipt',
       'travel_date', 'travel_time', 'travel_from', 'travel_to', 'car_type',
       'max_capacity'],
      dtype='object')

## <b> Data Description </b>

### <b>Nairobi Transport Data.csv (zipped)</b> is the dataset of tickets purchased from Mobiticket for the 14 routes from “up country” into Nairobi between 17 October 2017 and 20 April 2018. This dataset includes the variables: ride_id, seat_number, payment_method, payment_receipt, travel_date, travel_time, travel_from, travel_to, car_type, max_capacity.


### Uber Movement traffic data can be accessed [here](https://movement.uber.com). Data is available for Nairobi through June 2018. Uber Movement provided historic hourly travel time between any two points in Nairobi. Any tables that are extracted from the Uber Movement platform can be used in your model.

### Variables description:

* #### ride_id: unique ID of a vehicle on a specific route on a specific day and time.
* #### seat_number: seat assigned to ticket
* #### payment_method: method used by customer to purchase ticket from Mobiticket (cash or Mpesa)
* #### payment_receipt: unique id number for ticket purchased from Mobiticket
* #### travel_date: date of ride departure. (MM/DD/YYYY)
* #### travel_time: scheduled departure time of ride. Rides generally depart on time. (hh:mm)
* #### travel_from: town from which ride originated
* #### travel_to: destination of ride. All rides are to Nairobi.
* #### car_type: vehicle type (shuttle or bus)
* #### max_capacity: number of seats on the vehicle

In [11]:
#first create copy of dataset before working with so that we have original data untouched
df1 = df.copy()

## **Check columns and their unique values to understand what they contain.**

In [12]:
#check unique values for categorical columnns
categorical_columns = df1.select_dtypes(include=['object'])
for col in categorical_columns:
  print(f'Unique values in categorical columns {col} are: {categorical_columns[col].unique()[:8]}\n')

Unique values in categorical columns seat_number are: ['15A' '14A' '8B' '19A' '11A' '18B' '25' '21B']

Unique values in categorical columns payment_method are: ['Mpesa' 'Cash']

Unique values in categorical columns payment_receipt are: ['UZUEHCBUSO' 'TIHLBUSGTE' 'EQX8Q5G19O' 'SGP18CL0ME' 'BM97HFRGL9'
 'B6PBDU30IZ' 'MZHGDGS6QZ' 'MYVTYFNXDZ']

Unique values in categorical columns travel_date are: ['17-10-17' '19-11-17' '26-11-17' '27-11-17' '20-04-18' '19-04-18'
 '18-04-18' '17-04-18']

Unique values in categorical columns travel_time are: ['7:15' '7:12' '7:05' '7:10' '7:09' '5:10' '9:50' '7:06']

Unique values in categorical columns travel_from are: ['Migori' 'Keroka' 'Homa Bay' 'Kisii' 'Keumbu' 'Rongo' 'Kijauri' 'Oyugis']

Unique values in categorical columns travel_to are: ['Nairobi']

Unique values in categorical columns car_type are: ['Bus' 'shuttle']



In [13]:
#check unique values for numerical columns
numerical_columns = df1._get_numeric_data()
for col in numerical_columns:
  print(f'Unique values in numerical columns {col} are: {numerical_columns[col].unique()[:8]}\n')

Unique values in numerical columns ride_id are: [1442 5437 5710 5777 5778 5781 5784 5789]

Unique values in numerical columns max_capacity are: [49 11]



In [14]:
#check count of unqiue ride_id
len(df1['ride_id'].unique())

6249

## **Handling missing values.**

In [15]:
#check null values
df1.isnull().sum()

ride_id            0
seat_number        0
payment_method     0
payment_receipt    0
travel_date        0
travel_time        0
travel_from        0
travel_to          0
car_type           0
max_capacity       0
dtype: int64

As we can see there are no null values in our dataset.

## **Finding Target variable.**
We are not given the targte varibale, so we have to find it.
I am using count of ride_id to find our target variable.

In [16]:
labels = df.groupby(['ride_id']).seat_number.count().rename('number_of_tickets').reset_index()
labels.head()

ride_id  number_of_tickets
0     1442                  1
1     5437                  1
2     5710                  1
3     5777                  5
4     5778                 31

## **Add target variable into main dataframe.**

In [17]:
#merge two datasets on 'ride_id' columns
df1 = pd.merge(df1, labels, how ='left', on = 'ride_id')

In [18]:
df1

ride_id seat_number payment_method payment_receipt travel_date  \
0         1442         15A          Mpesa      UZUEHCBUSO    17-10-17   
1         5437         14A          Mpesa      TIHLBUSGTE    19-11-17   
2         5710          8B          Mpesa      EQX8Q5G19O    26-11-17   
3         5777         19A          Mpesa      SGP18CL0ME    27-11-17   
4         5778         11A          Mpesa      BM97HFRGL9    27-11-17   
...        ...         ...            ...             ...         ...   
51640    13826          9B          Mpesa      8V2XDDZR6V    20-04-18   
51641    13809         18A          Mpesa      4PEBSVJSNK    20-04-18   
51642    13809         17A          Mpesa      LVN64LZDNN    20-04-18   
51643    13796         16B          Mpesa      REYBSKTYWN    20-04-18   
51644    14304           7          Mpesa      AQN7FBUSGP    14-11-17   

      travel_time travel_from travel_to car_type  max_capacity  \
0            7:15      Migori   Nairobi      Bus            49   
1            7:12      Migori   Nairobi      Bus            49   
2            7:05      Keroka   Nairobi      Bus            49   
3            7:10    Homa Bay   Nairobi      Bus            49   
4            7:12      Migori   Nairobi      Bus            49   
...           ...         ...       ...      ...           ...   
51640        8:00      Awendo   Nairobi      Bus            49   
51641        8:00      Migori   Nairobi      Bus            49   
51642        8:00      Migori   Nairobi      Bus            49   
51643        7:08      Awendo   Nairobi      Bus            49   
51644        5:10       Kisii   Nairobi      Bus            49   

       number_of_tickets  
0                      1  
1                      1  
2                      1  
3                      5  
4                     31  
...                  ...  
51640                  1  
51641                 29  
51642                 29  
51643                 27  
51644                  1  

[51645 rows x 11 columns]

In [19]:
#check for duplicate rows
df1.duplicated().value_counts()

False    51645
dtype: int64

As we can see there is no duplicate row in our dataset.

## **Converting some columns to proper dtype and format.**

In [20]:
#as travel_date is in object dtype so we will convert it into datetime
df1['travel_date'] = pd.to_datetime(df['travel_date'],format = '%d-%m-%y')

In [21]:
#first convert travel_time from object to datetime
df1['travel_time'] = pd.to_datetime(df1['travel_time'])

In [22]:
#now convert travel_time to time format
df1['travel_time'] = df1['travel_time'].dt.time

In [27]:
#splitting travel_date column into day,month and year

def split_date(df1):
  '''This function takes dataframe as an argument and extract the different date features of the date variable and return updated dataset'''
  df1['day_of_month'] = df1['travel_date'].dt.day
  df1['month'] = df1['travel_date'].dt.month
  df1['year'] = df1['travel_date'].dt.year
  df1['day_of_week'] = df1['travel_date'].dt.dayofweek
  df1['is_weekend'] = df1['day_of_week'].apply(lambda x : 1 if x in[5,6] else 0)
  return df1

In [29]:
df1 = split_date(df1)

In [30]:
df1.head()

ride_id seat_number payment_method payment_receipt travel_date travel_time  \
0     1442         15A          Mpesa      UZUEHCBUSO  2017-10-17    07:15:00   
1     5437         14A          Mpesa      TIHLBUSGTE  2017-11-19    07:12:00   
2     5710          8B          Mpesa      EQX8Q5G19O  2017-11-26    07:05:00   
3     5777         19A          Mpesa      SGP18CL0ME  2017-11-27    07:10:00   
4     5778         11A          Mpesa      BM97HFRGL9  2017-11-27    07:12:00   

  travel_from travel_to car_type  max_capacity  number_of_tickets  \
0      Migori   Nairobi      Bus            49                  1   
1      Migori   Nairobi      Bus            49                  1   
2      Keroka   Nairobi      Bus            49                  1   
3    Homa Bay   Nairobi      Bus            49                  5   
4      Migori   Nairobi      Bus            49                 31   

   day_of_month  month  year  day_of_week  is_weekend  
0            17     10  2017            1           0  
1            19     11  2017            6           1  
2            26     11  2017            6           1  
3            27     11  2017            0           0  
4            27     11  2017            0           0

## **EDA**

In [66]:
#checking from which place people travel most
data = df1['travel_from'].value_counts()
fig = px.bar(data)
fig.show()